pzp/processing/merge_by_form_factor.py

```diff
@@ -0,0 +1,229 @@
+                mergeWithGeom = None
+                max = None
+                selFeat = QgsFeature()
+
+                # use prepared geometries for faster intersection tests
+                engine = QgsGeometry.createGeometryEngine(geom2Eliminate.constGet())
+                engine.prepareGeometry()
+
+                while fit.nextFeature(selFeat):
+                    if feedback.isCanceled():
+                        break
+
+                    selGeom = selFeat.geometry()
+
+                    if engine.intersects(selGeom.constGet()):
+                        # We have a candidate
+                        iGeom = geom2Eliminate.intersection(selGeom)
+
+                        # We need a common boundary in order to merge
+                        if not iGeom:
+                            continue
+
+                        selValue = None
+                        if mode == self.MODE_BOUNDARY:
+                            selValue = iGeom.length()
+
+                        elif mode == self.MODE_LARGEST_AREA:
+                            selValue = selGeom.area()
+
+                        elif mode == self.MODE_SMALLEST_AREA:
+                            selValue = selGeom.area() * -1
+
+                        else:
+                            raise QgsProcessingException(
+                                self.tr("Invalid value '{0}' for parameter '{1}'").format(mode, self.MODE)
+                            )
+
+                        if selValue is None:
+                            # No candidate found
+                            continue
+
+                        useThis = False
+                        if max is None:
+                            max = selValue
+                            useThis = True
+                        elif selValue > max:
+                            max = selValue
+                            useThis = True
+
+                        if useThis:
+                            mergeWithFid = selFeat.id()
+                            mergeWithGeom = QgsGeometry(selGeom)
+                # End while fit
+
+                if mergeWithFid is None:
+                    featNotEliminated.append(feat)
+                    continue
+
+                # A successful candidate
+                newGeom = mergeWithGeom.combine(geom2Eliminate)
+
+                if processLayer.changeGeometry(mergeWithFid, newGeom):
+                    madeProgress = True
+                else:
+                    raise QgsProcessingException(
+                        self.tr("Could not replace geometry of feature with id {0}").format(mergeWithFid)
+                    )
+
+                start = start + add
+                feedback.setProgress(start)
+
+            # End for featToEliminate
+
+            featToEliminate = featNotEliminated
+
+        # End while
+        if not processLayer.commitChanges():
+            raise QgsProcessingException(self.tr("Could not commit changes"))
```

Comment:
I think we should call `processLayer.rollBack()` if we couldn't commit, specially because we'll use the dataProvider to add features in line 216, so we should have a clean layer buffer status before that line.

Reply:
Ill add the rollback. We rise an exception so it will not execute the line 216